src/realm/object-store/sync/app_backing_store.hpp

```diff
@@ -0,0 +1,143 @@
+
+    // Gets the currently active user.
+    virtual std::shared_ptr<SyncUser> get_current_user() const = 0;
+
+    // Log out a given user
+    virtual void log_out_user(const SyncUser& user) = 0;
+
+    // Sets the currently active user.
+    virtual void set_current_user(const std::string& user_id) = 0;
+
+    // Removes a user
+    virtual void remove_user(const std::string& user_id) = 0;
+
+    // Permanently deletes a user.
+    virtual void delete_user(const std::string& user_id) = 0;
+
+    // Destroy all users persisted state and mark oustanding User instances as Removed
+    // clean up persisted state.
+    virtual void reset_for_testing() = 0;
+
+    // Called on start up after construction.
+    // The benefit being that `shared_from_this()` will work here.
+    virtual void initialize() = 0;
+
+    // FIXME: this is an implementation detail leak and doesn't belong in this API
+    // FIXME: consider abstracting it to something called `on_manual_client_reset()`
+    // Immediately run file actions for a single Realm at a given original path.
+    // Returns whether or not a file action was successfully executed for the specified Realm.
+    // Preconditions: all references to the Realm at the given path must have already been invalidated.
+    // The metadata and file management subsystems must also have already been configured.
+    virtual bool immediately_run_file_actions(const std::string& original_name) = 0;
+
+    // If the metadata manager is configured, perform an update. Returns `true` if the code was run.
+    virtual bool perform_metadata_update(util::FunctionRef<void(SyncMetadataManager&)> update_function) const = 0;
+
+    // Get the default path for a Realm for the given SyncUser.
+    // The default value is `<rootDir>/<appId>/<userId>/<partitionValue>.realm`.
+    // If the file cannot be created at this location, for example due to path length restrictions,
+    // this function may pass back `<rootDir>/<hashedFileName>.realm`
+    // The `user` is required.
+    // If partition_value is empty, FLX sync is requested
+    // otherwise this is for a PBS Realm and the string
+    // is a BSON formatted value.
+    virtual std::string path_for_realm(std::shared_ptr<SyncUser> user,
+                                       std::optional<std::string> custom_file_name = std::nullopt,
+                                       std::optional<std::string> partition_value = std::nullopt) const = 0;
+
+    // FIXME: see if this can be removed
+    // Get the path of the recovery directory for backed-up or recovered Realms.
+    virtual std::string
+    recovery_directory_path(std::optional<std::string> const& custom_dir_name = std::nullopt) const = 0;
+
+    // Get the app metadata for the active app.
+    virtual std::optional<SyncAppMetadata> app_metadata() const = 0;
+
+    // A pointer back to the parent App that this backing store supports.
+    std::weak_ptr<App> app() const
+    {
+        return m_parent_app;
+    }
```

Comment:
This relationship seems wrong. SyncUser should probably have a reference to its owning App rather than getting there indirectly (which is a preexisting problem with the class relationships). Anything outside of the Sync types and App touching the backing store directly seems incorrect.